geopandas/tools/overlay.py

```diff
@@ -176,3 +177,158 @@ def overlay(df1, df2, how, use_sindex=True):
+    dfsym.geometry = dfsym.new_g
+    dfsym = dfsym.loc[dfsym.geometry.is_empty==False].copy()
+    dfsym.drop(['bbox', 'sidx', 'idx', 'idx1','idx2', 'new_g'], axis=1, inplace=True)
+    dfsym = dfsym.reset_index(drop=True)
+    return dfsym
+
+def overlay(df1, df2, how='intersection', make_valid=True, reproject=True, use_sindex=None, **kwargs):
+    """Perform spatial overlay between two polygons.
+
+    Currently only supports data GeoDataFrames with polygons.
+    Implements several methods that are all effectively subsets of
+    the union.
+
+    Parameters
+    ----------
+    df1 : GeoDataFrame with MultiPolygon or Polygon geometry column
+    df2 : GeoDataFrame with MultiPolygon or Polygon geometry column
+    how : string
+        Method of spatial overlay: 'intersection', 'union',
+        'identity', 'symmetric_difference' or 'difference'.
+    reproject : boolean, default True
+        If GeoDataFrames do not have same projection, reproject
+        df2 to same projection of df1 before performing overlay
+
+    Returns
+    -------
+    df : GeoDataFrame
+        GeoDataFrame with new set of polygons and attributes
+        resulting from the overlay
+
+    """
+    if use_sindex is not None:
+        print('use_sindex is deprecated. If you are trying to use the old "overlay" function use "overlay_slow".')
+
+    # Allowed operations
+    allowed_hows = [
+        'intersection',
+        'union',
+        'identity',
+        'symmetric_difference',
+        'difference',  # aka erase
+    ]
+    # Error Messages
+    if how not in allowed_hows:
+        raise ValueError("`how` was \"%s\" but is expected to be in %s" % \
+            (how, allowed_hows))
+
+    if isinstance(df1, GeoSeries) or isinstance(df2, GeoSeries):
+        raise NotImplementedError("overlay currently only implemented for GeoDataFrames")
+
+    if (df1.geom_type.apply(lambda x: x in ['Polygon', 'MultiPolygon']).sum()!=len(df1.index) or 
+        df2.geom_type.apply(lambda x: x in ['Polygon', 'MultiPolygon']).sum()!=len(df2.index)):
+        raise TypeError("overlay only takes GeoDataFrames with (multi)polygon geometries") 
+
+    # Computations
+    df1 = df1.copy()
+    df2 = df2.copy()
+    df1['geometry'] = df1.geometry.buffer(0)
+    df2['geometry'] = df2.geometry.buffer(0)
+    if df1.crs!=df2.crs and reproject:
+        warnings.warn('Data has different projections.', UserWarning)
+        warnings.warn('Converted data to projection of first GeoPandas DataFrame', UserWarning)
+        df2.to_crs(crs=df1.crs, inplace=True)
+    if how=='intersection':
+        return overlay_intersection(df1, df2)
+    elif how=='difference':
+        return overlay_difference(df1, df2)
+    elif how=='symmetric_difference':
+        return overlay_symmetric_diff(df1, df2)
+    elif how=='union':
+        dfinter = overlay(df1, df2, how='intersection')
+        dfsym = overlay(df1, df2, how='symmetric_difference')
```

Comment:
I would here use the helper functions directly instead of the main `overlay` function. This avoids doing the checks in the beginning (including the 0 buffers) multiple times.